### ISSUE TITLE
False positive VUID-vkQueueSubmit-fence-00063 with fence and signal semaphore

### ISSUE THREAD

**Issue by jhs67** [KhronosGroup/Vulkan-ValidationLayers#4271]
Sometimes I will get incorrect validation errors from VUID-vkQueueSubmit-fence-00063. I call vkWaitForFences on a single fence that was previously used in a call to vkQueueSubmit. I immediately call vkResetFences. Then sometime later I reuse the fence in another call to vkQueueSubmit and the validation warning is sometimes triggered.

I believe the problem originates when there is both a fence and a signal semapore on the queue submission.

**Valid Usage ID**
VUID-vkQueueSubmit-fence-00063: Validation Error: [ VUID-vkQueueSubmit-fence-00063 ] Object 0: handle = 0xf443490000000006, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x6f35a2b3 | vkQueueSubmit(): VkFence 0xf443490000000006[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-fence-00063)

**Environment:**
 - OS: Linux
 - GPU: NVIDIA Corporation GP106 [GeForce GTX 1060 3GB] (rev a1)
 - SDK or header version if building from repo: 1.3.216.0
 - Options enabled (synchronization, best practices, etc.): VK_LAYER_KHRONOS_validation

This occurs on the graphics queue submission and I include a fence and a semephore that I latter pass to vkQueuePresentKHR. In a separate thread I call vkWaitForFences on the fence.

The rest is speculation: I think the presentation library calls vkWaitSemaphores in another thread. When the queue finishes running, both threads are freed to run. There is a race to call QUEUE_STATE::Retire. The presentation thread wins that race and begins processing the submission. My thread returns and immediatly calls vkFenceReset that sets the FENCE_STATE::state_ to FENCE_UNSIGNALED. The presentation thread continues processing the submission and finishes by calling FENCE_STATE::Retire, which promptly sets FENCE_STATE::state_ back to FENCE_RETIRED (incorrectly).

I'm not sure how all this works, but maybe you could check the FENCE_STATE::queue_ to see if the fence still needs to be reset?

**Comment by jeremyg-lunarg** [comment]
I agree there's a possible race condition there, need to figure out a good fix for it.
